iOS v4 booking (#112)
MapsPeople/docsite: docs/06-Working with Data/02-Booking/index.mdx

1. Call `getBookableLocations()` on the `MPBookingService` instance.
1. Assign the returned locations to your `bookableLocations` property.

```swift
override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)

    self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Auth", style: .plain, target: self, action: #selector(openAuthConfig))
    self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Login", style: .plain, target: self, action: #selector(openLogin))

    Task {
        await refresh()
    }
}
```

add a refresh function:

```swift
func refresh() async {
    let startTime = Date()
    let endTime = startTime.advanced(by: 60*60)
    let bookableQuery = MPBookableQuery(startTime: startTime, endTime: endTime)

    let bookingService = MPMapsIndoors.shared.bookingService

    if let locations = try? await bookingService.locationsConfiguredForBooking(bookableQuery) {
        locationsConfiguredForBooking = locations
        tableView.reloadData()
    }

    if let locations = try? await bookingService.bookableLocationsUsing(query: bookableQuery) {
        bookableLocations = locations
        tableView.reloadData()
    }
}
```

> In `numberOfSections(in tableView:)` add:

Reviewer: ```suggestion
In `numberOfSections(in:)` add:
```

```swift
override func numberOfSections(in tableView: UITableView) -> Int {
    return 2
}
```

> In `tableView(:numberOfRowsInSection section:)`, return the size of `bookableLocations`.

Reviewer: ```suggestion
In `tableView(_:numberOfRowsInSection:) `, return the size of `bookableLocations`.
```

```swift
override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    if section == 0 {
        return bookableLocations.count
    } else {
        return locationsConfiguredForBooking.count
    }
}
```

> In `tableView(:cellForRowAt indexPath:)`, create a `UITableViewCell` with the current Locations name as the text for the label.

Reviewer: ```suggestion
In `tableView(_:cellForRowAt:)`, create a `UITableViewCell` with the current Locations name as the text for the label.
```

```swift
override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = UITableViewCell()
    var list = locationsConfiguredForBooking
    if indexPath.section == 0 {
        list = bookableLocations
    }
    let location = list[indexPath.row]
    cell.textLabel?.text = location.name
    return cell
}
```

> In `tableView(:didSelectRowAt indexPath:)`, get the relevant `MPLocation` for the `indexPath`. Initialise a `BookingsController` which we will implement next. Assign the selected Location to a `bookableLocation` property on `BookingsController` and push the controller to the navigation stack.

Reviewer: ```suggestion
In `tableView(_:didSelectRowAt:)`, get the relevant `MPLocation` for the `indexPath`. Initialise a `BookingsController` which we will implement next. Assign the selected Location to a `bookableLocation` property on `BookingsController` and push the controller to the navigation stack.
```

```swift
private var bookings = [MPBooking]()
```

In your `viewDidLoad()` method, initialise a `UIBarButtonItem` with the title `Book`targeting `newBooking` which we will create later. Add the button to the `navigationItem`.

```swift
let button = UIBarButtonItem(title: "Book", style: .plain, target: self, action: #selector(newBooking))
self.navigationItem.rightBarButtonItem = button
```

Also in your `viewDidLoad()` method, initialise a `MPBookingsQuery`with the `MPLocation` stored in `bookableLocation` and a timespan, in this example 24 hours starting one hour ago.

```swift
let bookingsQuery = MPBookingsQuery()
bookingsQuery.location = bookableLocation
bookingsQuery.startTime = Date().advanced(by: -60*60)
bookingsQuery.endTime = bookingsQuery.startTime?.advanced(by: 24*60*60)
```

Lastly in your `viewDidLoad()` method, call `bookingsUsing(query: bookingsQuery)`with the `bookingsQuery` we just created. Store the returned Bookings in our `bookings` property.

```swift
Task {
    if let bookings = try await MPMapsIndoors.shared.bookingService.bookingsUsing(query: bookingsQuery) {
        self.bookings = bookings
        tableView.reloadData()
    }
}
```

> In `numberOfSections(in tableView:)`, return 1 since we only need one section.

Reviewer: ```suggestion
In `numberOfSections(in:)`, return 1 since we only need one section.
```

```swift
override func numberOfSections(in tableView: UITableView) -> Int {
    return 1
}
```

> In `tableView(:numberOfRowsInSection section:)`, return the size of `bookings`.

Reviewer: ```suggestion
In `tableView(_:numberOfRowsInSection:)`, return the size of `bookings`.
```

```swift
override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return bookings.count
}
```

> In `tableView(:cellForRowAt indexPath:)`, create a `UITableViewCell` with the current Booking title as the text for the label.

Reviewer: ```suggestion
In `tableView(_:cellForRowAt:)`, create a `UITableViewCell` with the current Booking title as the text for the label.
```

```swift
override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = UITableViewCell()
    let booking = bookings[indexPath.row]
    cell.textLabel?.text = booking.title
    return cell
}
```

> In `tableView(:didSelectRowAt indexPath:)`, get the relevant `MPBooking` for the `indexPath` and call `editBooking()` with that Booking.

Reviewer: ```suggestion
In `tableView(_:didSelectRowAt:)`, get the relevant `MPBooking` for the `indexPath` and call `editBooking()` with that Booking.
```

}
```

Create another Objective-C exposed method `cancel()` which will be use by our `UIBarButtonItem` inserted in `viewDidLoad()`. In the `cancel()` mehod, call `cancel(booking)` on the `MPBookingService` instance with our Booking object as input. If all goes well and we have a Booking returned in the block, we assign this new Booking to our `booking` propery and refresh our views. If not, we assume that we have an error, and show this in an alert controller.

```swift
@objc private func cancel() async {
        do {
            if (try await MPMapsIndoors.shared.bookingService.cancelBooking(booking)) != nil {
                let alert = UIAlertController(title: "Booking was cancelled!", message: "Booking was successfully cancelled!", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                present(alert, animated: true)
            }
        } catch {
            let alert = UIAlertController(title: "Ooops!", message: "\(error)", preferredStyle: .alert)
            alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
            present(alert, animated: true)
        }
    }
```

> In `numberOfSections(in tableView:)`, return 1 since we only need one section.

Reviewer: ```suggestion
In `numberOfSections(in:)`, return 1 since we only need one section.
```

```swift
override func numberOfSections(in tableView: UITableView) -> Int {
    return 1
}
```

> In `tableView(:numberOfRowsInSection section:)`, return the value of `BookingRow.count.rawValue`.

Reviewer: ```suggestion
In `tableView(_:numberOfRowsInSection:)`, return the value of `BookingRow.count.rawValue`.
```

```swift
override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
    return BookingRow.count.rawValue
}
```

> In `tableView(:cellForRowAt indexPath:)`, create a `UITableViewCell` and create a switch control structure by initialising a `BookingRow` enum value from `indexPath.row`. Based on the cases, populate the `textLabel` with `title`, `bookingDescription`, `startTime`, `endTime` and `bookingId` from your `MPBooking` instance.

Reviewer: ```suggestion
In `tableView(_:cellForRowAt:)`, create a `UITableViewCell` and create a switch control structure by initialising a `BookingRow` enum value from `indexPath.row`. Based on the cases, populate the `textLabel` with `title`, `bookingDescription`, `startTime`, `endTime` and `bookingId` from your `MPBooking` instance.
```

```swift
override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    let cell = UITableViewCell.init(style: .subtitle, reuseIdentifier: nil)
    switch BookingRow.init(rawValue: indexPath.row)  {
    case .title:
        cell.textLabel?.text = booking.title ?? ""
        cell.detailTextLabel?.text = "Title"
    case .description:
        cell.textLabel?.text = booking.bookingDescription ?? ""
        cell.detailTextLabel?.text = "Description"
    case .start:
        cell.textLabel?.text = "\(booking.startTime ?? Date())"
        cell.detailTextLabel?.text = "Start time"
    case .end:
        cell.textLabel?.text = "\(booking.endTime ?? Date().addingTimeInterval(60*60))"
        cell.detailTextLabel?.text = "End time"
    case .id:
        cell.textLabel?.text = booking.bookingId ?? ""
        cell.detailTextLabel?.text = "Booking id"
    default : ()
    }
    return cell
}
```

> In `tableView(:cellForRowAt indexPath:)`, create a switch control structure again by initialising a `BookingRow` enum value from `indexPath.row`. Based on the cases, either initialise and present `FieldEditController` or `DatePickerController` which we will implement next.

Reviewer: ```suggestion
In `tableView(_:cellForRowAt:)`, create a switch control structure again by initialising a `BookingRow` enum value from `indexPath.row`. Based on the cases, either initialise and present `FieldEditController` or `DatePickerController` which we will implement next.
```

</Tabs>

</TabItem>

<TabItem value="ios-v4" label="iOS v4">

This guide covers the different aspects of _Booking_ in the MapsIndoors iOS SDK. The concept of Booking in MapsIndoors implies that specific Locations in your MapsIndoors dataset is treated as Bookable resources. Typical bookable resources could be meeting rooms and workspaces.

A MapsIndoors dataset can only have bookable resources if an integration with a Booking provider exists. Current examples of Booking providers are _Google Calendar_ and _Microsoft Office 365_. These providers and more can be added and integrated to your MapsIndoors project by request.

The central service in the SDK managing Bookings is the Booking Service.

```swift

> MPMapsIndoors.shared.bookingService.bookingsUsing(query: MPBookingsQuery)

Reviewer: ```suggestion
MPMapsIndoors.shared.bookingService
```

> ## Bookable Locations for iOS

Reviewer: No need for "iOS" in the headline
```suggestion
## Bookable Locations
```

> To determine whether or not a Location is bookable can be looked up using the `MPMapsIndoors.shared.bookingService.bookingsUsing(query: MPBookingsQuery)` method. Below is an example of querying for bookable Locations:

Reviewer: ```suggestion
To determine whether or not a Location is bookable can be looked up using the `MPMapsIndoors.shared.bookingService. bookableLocationsUsing(query: MPBookableQuery)` method. Below is an example of querying for bookable Locations:
```

```swift
private var bookings = [MPBooking]()

let bookingsQuery = MPBookingsQuery.init()
bookingsQuery.startTime = Date.init().advanced(by: -60*60)
bookingsQuery.endTime = bookingsQuery.startTime?.advanced(by: 24*60*60)
    
Task {
    if let bookings = try await MPMapsIndoors.shared.bookingService.bookingsUsing(query: bookingsQuery) {
        self.bookings = bookings
    }

> }

Reviewer: There is a big difference between `bookableLocations` and `bookings`.

Please revert this example to be about `bookableLocations` and use `bookableLocationsUsing(query:)`

## Bookings for iOS

A Booking is a timeboxed event model referring to the resource being booked and the users participating in the booked event.

### Listing Bookings for a Location for iOS

Before trying to book a Location for a given time, it is convenient to know in advance whether or not the Location is already booked for the given time.

It is possible to get a list of bookings using the `MPBookingService.getBookingsUsing()` method.

```swift
let bookingService = MPBookingService.sharedInstance()
let bookingsQuery = MPBookingsQuery.init()

bookingsQuery.location = myRoomLocation
bookingsQuery.startTime = Date.init().advanced(by: -60*60)
bookingsQuery.endTime = bookingsQuery.startTime?.advanced(by: 24*60*60)

bookingService.getBookingsUsing(bookingsQuery) { (bookings, error) in
}
```

> The above example creates a query for bookings that exists for a location with timespan between 1 hour ago and 24 hours ahead.

Reviewer: This is still needed as per above comment on difference between `bookableLocations` and `bookings`

> ### Performing a Booking of a Location for iOS

Reviewer: No need for "iOS" in the headline
```suggestion
### Performing a Booking of a Location
```

>     let booking = MPBooking.init()

Reviewer: ```suggestion
    let booking = MPBooking()
```

>     booking.startTime = Date.init()

Reviewer: ```suggestion
    booking.startTime = Date()
```

Depending on the Booking provider, the participants will receive invites for an event created by this Booking request.

> ### Cancelling a Booking of a Location for iOS

Reviewer: No need for "iOS" in the headline
```suggestion
### Cancelling a Booking of a Location
```

It is possible to cancel a created Booking using the `MPBookingService.cancel()` method which takes an existing Booking object as input.

```swift
do {
    if (try await MPMapsIndoors.shared.bookingService.cancelBooking(booking)) != nil {
        let alert = UIAlertController(title: "Booking was cancelled!", message: "Booking was successfully cancelled!", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        present(alert, animated: true)
    }
} catch {
    let alert = UIAlertController(title: "Ooops!", message: "\(error)", preferredStyle: .alert)
    alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
    present(alert, animated: true)
}
```

> ## Booking in Practice for iOS

Reviewer: No need for "iOS" in the headline
```suggestion
## Booking in Practice
```

```swift
override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)

>     self.navigationItem.rightBarButtonItem = UIBarButtonItem.init(title: "Auth", style: .plain, target: self, action: #selector(openAuthConfig))
>     self.navigationItem.rightBarButtonItem = UIBarButtonItem.init(title: "Login", style: .plain, target: self, action: #selector(openLogin))

Reviewer: ```suggestion
    self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Auth", style: .plain, target: self, action: #selector(openAuthConfig))
    self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Login", style: .plain, target: self, action: #selector(openLogin))
```

```swift
override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

>     let cell = UITableViewCell.init()

Reviewer: ```suggestion
    let cell = UITableViewCell()
```

```swift
override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {

>     let bookingsVC = BookingsController.init()

Reviewer: ```suggestion
    let bookingsVC = BookingsController()
```

var booking = MPBooking.init()
```

Create a method `updateButtons()` that will either place a `book button` if there is no `bookingId` on the Booking, which means it was created locally, or place a `cancel button` if a `bookingId` exist for the Booking, which means it was selected from a list of Bookings fetched from the `MPBookingService`.

```swift
private func updateButtons() {
    if booking.bookingId != nil {

>         self.navigationItem.rightBarButtonItem = UIBarButtonItem.init(title: "Cancel Booking", style: .plain, target: self, action: #selector(cancel))

Reviewer: ```suggestion
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Cancel Booking", style: .plain, target: self, action: #selector(cancel))
```

    } else {

>         self.navigationItem.rightBarButtonItem = UIBarButtonItem.init(title: "Create Booking", style: .plain, target: self, action: #selector(book))

Reviewer: ```suggestion
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Create Booking", style: .plain, target: self, action: #selector(book))
```
